Factorial using 2-partition algorithm

David5i6 opened the pull request:
fixes #5972 

Implementation of factorial using 2 partition algorithm

When using **TimeProfiler** the original recursive implementation became  so slow while it runs fast with non recursive versions.

Also this algorithm is almost twice fast than the original without usage of prime numbers.

David5i6 committed:
Factorial using 2-partition algorithm

guillep commented:
Without much to say, I wonder what the people around polymath say.
@SergeStinckwich , @olekscode , if you have 10 minutes to give an opinion during these days, it would be great :)

Ducasse commented:
```
^ self < 0
 				ifTrue: [ self error: 'Not valid for negative integers' ]
 				ifFalse: [ self > 0
 						ifTrue: [ ^ self ]
 						ifFalse: [ ^ 1 ] ] ].
```

this code has too many returns :)

```
^ self < 0
 				ifTrue: [ self error: 'Not valid for negative integers' ]
 				ifFalse: [ self > 0
 						ifTrue: [ self ]
 						ifFalse: [  1 ] ] ].
```

Ducasse commented:
can you rename factorial as slow factorial and add a comment stating that we know that is is slow but that we keep it for the fun.

David5i6 committed:
factorial renamed as slowfactorial, added new factorial with improved speed and less resource consumption.
Added tests

David5i6 commented:
I added some tests, comments and moved original to slowfactorial.

I've seen other comments that point to let original as is, and put new as fastfactorial....

olekscode commented on src/Kernel-Tests/IntegerTest.class.st:
```diff
@@ -1101,6 +1107,18 @@ IntegerTest >> testRounding [
+{ #category : #'tests - bitLogic' }
+IntegerTest >> testSlowFactorial [
+
+	"Check some know values"
```
"Check some **known** values"

olekscode commented on src/Kernel-Tests/IntegerTest.class.st:
```diff
@@ -382,6 +382,12 @@ IntegerTest >> testEven [
 ]
 
+{ #category : #'tests - bitLogic' }
+IntegerTest >> testFactorial [	
```
I think it is better to confider `factorial` as the default implementation. So instead of asserting if `factorial` is equal to `slowFactorial`, I think it is better to have a normal test here, and then in `testSlowFactorial`, we can assert whether `slowFactorial` is equal to `factorial`. This way we would reverse the dependency and it will be `slowFactorial` that depends on `factorial` and not the other way around. It is good because `slowFactorial` could be removed or renamed in the future.

Ducasse commented:
could you rename slowfactorial as slowFactorial?

pavel-krivanek commented:
it would be more newcomers friendly if it would be `factorialSlow` because then these methods would be next to each other...

David5i6 committed:
renamed slowfactorial to slowFactorial.
Test of factorial has no dependency of slowFactorial.

David5i6 committed:
Correction on slowfactorial executable comment (no autorefactored).

Also small improvement in slowFactorial, chaged order of the if, so for numbers greater than 0 there will be less if checks. Teorically faster, but in practice almost unnoticeable.

David5i6 commented:
After reading comments I've made some small changes. Renamed *slowfactorial" to "slowFactorial".
Tests of factorial does not have dependency of slowFactorial.
Also a small (unnoticeable) improvement in slowFactorial changing order of the ifs.

guillep commented:
I think this is good enough. We can always improve later.

Thanks @David5i6 !
